### PR TITLE
Add MongoDB Spanstore

### PR DESCRIPTION
**Pull request by maz**
Add support for storing traces in MongoDB using the new SpanStore API.

### CODE REVIEW

File: zipkin-mongodb/src/main/scala/com/twitter/zipkin/mongodb/MongoDBSpanStoreFactory.scala [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.twitter.zipkin.mongodb
+
+import java.util.concurrent.TimeUnit
+
+import com.twitter.app.App
+import com.twitter.util.Duration
+import com.twitter.zipkin.storage.SpanStore
+import com.twitter.zipkin.storage.mongodb.MongoDBSpanStore
+
+trait MongoDBSpanStoreFactory{ self: App =>
```
Comment (sprsquish) [inline zipkin-mongodb/src/main/scala/com/twitter/zipkin/mongodb/MongoDBSpanStoreFactory.scala]: style: space between the trait and the brace

File: zipkin-mongodb/src/main/scala/com/twitter/zipkin/storage/mongodb/MongoDBSpanStore.scala [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+import com.twitter.zipkin.storage.mongodb.utils.{EnsureIndexes, Index}
+import com.twitter.zipkin.storage.{IndexedTraceId, SpanStore, TraceIdDuration}
+
+private[mongodb] object RegisterMongoDBSerializers {
+  private[this] lazy val registerSerializers = new Serializers {}.register()
+
+  def apply(): Unit = registerSerializers
+}
+
+private[mongodb] class Asyncifier {
```
Comment (sprsquish) [inline zipkin-mongodb/src/main/scala/com/twitter/zipkin/storage/mongodb/MongoDBSpanStore.scala]: extends com.twitter.util.Closable

File: zipkin-mongodb/src/main/scala/com/twitter/zipkin/storage/mongodb/MongoDBSpanStore.scala [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+import com.mongodb.casbah.Imports._
+import com.mongodb.casbah.commons.conversions.scala._
+import com.twitter.finagle.jsr166y.ForkJoinPool
+import com.twitter.util._
+import com.twitter.zipkin.common._
+import com.twitter.zipkin.gen
+import com.twitter.zipkin.storage.mongodb.utils.{EnsureIndexes, Index}
+import com.twitter.zipkin.storage.{IndexedTraceId, SpanStore, TraceIdDuration}
+
+private[mongodb] object RegisterMongoDBSerializers {
```
Comment (sprsquish) [inline zipkin-mongodb/src/main/scala/com/twitter/zipkin/storage/mongodb/MongoDBSpanStore.scala]: Package level scoping of visibility is the default. There's no need to put "[mongodb]" after "private"

File: zipkin-mongodb/src/main/scala/com/twitter/zipkin/storage/mongodb/MongoDBSpanStore.scala [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+    val obj = new MongoDBObject(dbobj)
+    IndexedTraceId(
+      obj.as[Long]("traceId"),
+      startTimeStampFromMongoObject(obj)
+    )
+  }
+
+  private[mongodb] def toByteArray(buffer: ByteBuffer): Array[Byte] = {
+    val arr = new Array[Byte](buffer.remaining())
+    val copy = buffer.slice() //get() modifies the ByteBuffer position. Slice makes a shallow copy so this isn't a problem
```
Comment (sprsquish) [inline zipkin-mongodb/src/main/scala/com/twitter/zipkin/storage/mongodb/MongoDBSpanStore.scala]: buffer.duplicate() is more idiomatic

File: zipkin-mongodb/src/main/scala/com/twitter/zipkin/storage/mongodb/MongoDBSpanStore.scala [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+  override def getSpansByTraceId(traceId: Long): Future[Seq[Span]] = optionalGetSpansByTraceId(traceId).map(_.get)
+
+  /**
+   * Get the trace ids for this annotation between the two timestamps. If value is also passed we expect
+   * both the annotation key and value to be present in index for a match to be returned.
+   * Only return maximum of limit trace ids from before the endTs.
+   */
+  override def getTraceIdsByAnnotation(serviceName: String, annotation: String, value: Option[ByteBuffer],
+                                       endTs: Long, limit: Int): Future[Seq[IndexedTraceId]] =
+    if (Seq(//The Spec says that these shouldn't be searchable
```
Comment (sprsquish) [inline zipkin-mongodb/src/main/scala/com/twitter/zipkin/storage/mongodb/MongoDBSpanStore.scala]: You can use the already existing list: https://github.com/twitter/zipkin/blob/82594bb081a829ad43c55962b73fdab977d70437/zipkin-common/src/main/scala/com/twitter/zipkin/Constants.scala#L32

File: project/Project.scala [hunk 6]
```diff
@@ -508,7 +521,7 @@ object Zipkin extends Build {
       twitterServer
     )
   ).dependsOn(
-    tracegen, web, anormDB, query,
+    tracegen, web, mongodb, query,
```
Comment (sprsquish) [inline project/Project.scala]: Revert this. zipkin-example should continue to use AnormDB as its default.

**Comment by sprsquish** [comment]
Thanks for this. I took one pass at it and will take another pass tomorrow. 
There isn't a written style guide, but the style here doesn't match the rest of the project very well.

**Comment by maz** [comment]
Do you mean style in terms of whitespace formatting or in terms of how the code itself is structured?

**Comment by sprsquish** [comment]
How it's structured. It's not a show stopper. I'll probably end up cleaning it up. It's on me to have a style guide in place and not simply expect people to know what I'm looking for :)

**Comment by hunglin** [comment]
@sprsquish , it got OutOfMemoryError from Travis CI build, is this a real issue or just a hiccup of Travis?  How to do it if I want to relaunch the CI build?

**Comment by sprsquish** [comment]
@hunglin I'll restart it. Travis is notoriously flaky for Zipkin so I don't put too much stock in it. It may be a few days before I'll have a chance to re-review.

**Comment by hunglin** [comment]
Thanks a lot!

**Comment by hunglin** [comment]
openjdk6 passed this time, maybe the old saying is true: third time's the charm

**Comment by hunglin** [comment]
one question, @sprsquish , it seems zipkin/sbt set -XX:MaxPermSize=1024m but Travis doesn't use it so the the MaxPermSize is 512M for ci build, that's probably the reason.

**Comment by sprsquish** [comment]
@hunglin good catch. I'll see if I can get Travis to use our script instead.

**Comment by hunglin** [comment]
it passed! 
celebrate

**Comment by codefromthecrypt** [comment]
This seems to be uncontested, and accidentally overlooked. I'll see if this is mergable still or not.

**Comment by codefromthecrypt** [comment]
@maz well.. better late then never? :) closing via 3875f2a7eb6cc7e2baa6568196e1f34706127805

File: zipkin-mongodb/src/main/scala/com/twitter/zipkin/mongodb/MongoDBSpanStoreFactory.scala (hunk 1 above)
Comment (qiufeng001) [reply zipkin-mongodb/src/main/scala/com/twitter/zipkin/mongodb/MongoDBSpanStoreFactory.scala]: Do you have a version with mongodb? Thanks

**Comment by codefromthecrypt** [comment]
> Do you have a version with mongodb? Thanks

once there was but it was removed years ago due to no maintenance. the code
from back then is irrelevant now because the model has changed also
language and frameworks.

**Comment by qiufeng001** [comment]
> Do you have a version with mongodb? Thanks
> once there was but it was removed years ago due to no maintenance. the code from back then is irrelevant now because the model has changed also language and frameworks.

No, but thank you very much.

**Pull request closed without merging**